x/twap/migrate_test.go

```diff
@@ -93,3 +93,189 @@ func (suite *TestSuite) TestTwapRecord_GeometricTwap_MarshalUnmarshal() {
+		}
+		return records
+	}
+
+	// concatenates all element slices into a single slice and
+	// returns the result.
+	merge := func(records [][]types.TwapRecord) []types.TwapRecord {
+		var result []types.TwapRecord
+		for _, r := range records {
+			r := r
+			result = append(result, r...)
+		}
+		return result
+	}
+
+	// returns three asset record with the given pool id and time, everything other parameter set to 1.
+	threeAssetRecordWithPoolIdAndTime := func(poolId uint64, time time.Time) []types.TwapRecord {
+		return newThreeAssetRecord(poolId, time, sdk.OneDec(), sdk.OneDec(), sdk.OneDec(), sdk.OneDec(), sdk.OneDec(), sdk.OneDec(), sdk.OneDec())
+	}
+
+	tests := map[string]struct {
+		expectError bool
+
+		preExistingRecords []types.TwapRecord
+
+		// pool id -> most recent records
+		expectedMostRecent map[uint64][]types.TwapRecord
+	}{
+		"one record, one pool": {
+			preExistingRecords: []types.TwapRecord{
+				baseRecord,
+			},
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				baseRecord.PoolId: updateGeometricAccum([]types.TwapRecord{baseRecord}),
+			},
+		},
+		"three records, one pool": {
+			preExistingRecords: threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime),
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				basePoolId: updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime)),
+			},
+		},
+		"two pools, one record each": {
+			preExistingRecords: []types.TwapRecord{
+				baseRecord,
+				withPoolId(baseRecord, baseRecord.PoolId+1),
+			},
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				baseRecord.PoolId:     updateGeometricAccum([]types.TwapRecord{baseRecord}),
+				baseRecord.PoolId + 1: updateGeometricAccum([]types.TwapRecord{withPoolId(baseRecord, baseRecord.PoolId+1)}),
+			},
+		},
+		"two pools, multiple records each at the same time": {
+			preExistingRecords: append(
+				threeAssetRecordWithPoolIdAndTime(baseRecord.PoolId, baseTime),
+				threeAssetRecordWithPoolIdAndTime(baseRecord.PoolId+1, baseTime)...),
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				baseRecord.PoolId:     updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime)),
+				baseRecord.PoolId + 1: updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId+1, baseTime)),
+			},
+		},
+		"two pools, multiple records each at different times": {
+			preExistingRecords: append(
+				threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime),
+				threeAssetRecordWithPoolIdAndTime(basePoolId+1, baseTime.Add(time.Second))...),
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				baseRecord.PoolId:     updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime)),
+				baseRecord.PoolId + 1: updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId+1, baseTime.Add(time.Second))),
+			},
+		},
+		"three pools, 2 at the same time, 1 is not, entries and different time for each": {
+			preExistingRecords: merge([][]types.TwapRecord{
```

Comment:
why not put in []types.TwapRecord directly instead of providing here [][] and using merge function?

Reply:
Well, each call to `threeAssetRecordWithPoolIdAndTime` returns three different []types.TwapRecord that must be merged into a single slice. What is the alternative? We could used `append` multiple times but it's not as clean